src/wazuh_modules/agent_upgrade/manager/wm_agent_upgrade_upgrades.c

```diff
@@ -0,0 +1,542 @@
+
+            w_mutex_unlock(&upgrade_threads_mutex);
+        }
+    }
+
+    // Destroy threads count mutex
+    w_mutex_destroy(&upgrade_threads_mutex);
+
+    // Destroy threads count condition variable
+    w_cond_destroy(&upgrade_threads_cond);
+
+    return NULL;
+}
+
+void* wm_agent_upgrade_start_upgrade(void *arg) {
+    wm_upgrade_args *upgrade_config = (wm_upgrade_args *)arg;
+
+    wm_manager_configs *config = upgrade_config->config;
+    wm_agent_task *agent_task = upgrade_config->agent_task;
+
+    do {
+        int error_code = WM_UPGRADE_SUCCESS;
+
+        if (error_code = wm_agent_upgrade_send_wpk_to_agent(agent_task, config), error_code == WM_UPGRADE_SUCCESS) {
+
+            if (WM_UPGRADE_UPGRADE == agent_task->task_info->command) {
+                wm_upgrade_task *upgrade_task = agent_task->task_info->task;
+
+                if (upgrade_task->custom_version && (wm_agent_upgrade_compare_versions(upgrade_task->custom_version, WM_UPGRADE_NEW_UPGRADE_MECHANISM) < 0)) {
+
+                    // Update task to "Legacy". The agent won't report the result of the upgrade task
+                    cJSON *status_request = cJSON_CreateArray();
+                    cJSON *status_response = cJSON_CreateArray();
+
+                    cJSON_AddItemToArray(status_request, wm_agent_upgrade_parse_task_module_request(WM_UPGRADE_AGENT_UPDATE_STATUS, agent_task->agent_info->agent_id, task_statuses[WM_TASK_LEGACY], NULL));
+                    wm_agent_upgrade_task_module_callback(status_response, status_request, NULL, NULL);
+                    wm_agent_upgrade_validate_task_status_message(cJSON_GetArrayItem(status_response, 0), NULL, NULL);
+
+                    cJSON_Delete(status_request);
+                    cJSON_Delete(status_response);
+                }
+            }
+        } else {
+
+            // Update task to "Failed"
+            cJSON *status_request = cJSON_CreateArray();
+            cJSON *status_response = cJSON_CreateArray();
+
+            cJSON_AddItemToArray(status_request, wm_agent_upgrade_parse_task_module_request(WM_UPGRADE_AGENT_UPDATE_STATUS, agent_task->agent_info->agent_id, task_statuses[WM_TASK_FAILED], upgrade_error_codes[error_code]));
+            wm_agent_upgrade_task_module_callback(status_response, status_request, NULL, NULL);
+            wm_agent_upgrade_validate_task_status_message(cJSON_GetArrayItem(status_response, 0), NULL, NULL);
+
+            cJSON_Delete(status_request);
+            cJSON_Delete(status_response);
+        }
+
+        wm_agent_upgrade_free_agent_task(agent_task);
+
+        struct timespec timeout = { .tv_sec = time(NULL) + 1 };
+        agent_task = queue_pop_ex_timedwait(upgrade_queue, &timeout);
+
+    } while (agent_task);
+
+    w_mutex_lock(&upgrade_threads_mutex);
+
+    upgrade_threads_count--;
+
+    // Notify execution finished
+    w_cond_signal(&upgrade_threads_cond);
+
+    w_mutex_unlock(&upgrade_threads_mutex);
```

Comment:
Since I already added a function to wait for an available thread let's add a method to release it:

```
void wm_agent_upgrade_release_thread() {
    w_mutex_lock(&upgrade_threads_mutex);

    upgrade_threads_count--;

    // Notify execution finished
    w_cond_signal(&upgrade_threads_cond);

    w_mutex_unlock(&upgrade_threads_mutex);
}
```

Reply:
Done!